File: src/atc/core/sentry.py
"""Sentry SDK initialisation and PII stripping for ATC backend."""

from __future__ import annotations

import logging
import re
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from atc.config import SentryConfig

logger = logging.getLogger(__name__)

# Patterns that look like PII — matched case-insensitively in event data.
_PII_KEY_PATTERNS = re.compile(
    r"(password|secret|token|api.?key|auth|credential|cookie|session.?id|"
    r"email|phone|ssn|credit.?card|card.?number)",
    re.IGNORECASE,
)

_REDACTED = "[Filtered]"


def _strip_pii(obj: Any) -> Any:
    """Recursively redact values whose keys look like PII."""
    if isinstance(obj, dict):
        return {
            k: (_REDACTED if _PII_KEY_PATTERNS.search(k) else _strip_pii(v))
            for k, v in obj.items()
        }
    if isinstance(obj, list):
        return [_strip_pii(item) for item in obj]
    return obj


def _before_send(event: dict[str, Any], hint: dict[str, Any]) -> dict[str, Any] | None:
    """Strip PII from Sentry events before they leave the process."""
    # Scrub request data
    if "request" in event:
        req = event["request"]
        if "headers" in req:
            req["headers"] = _strip_pii(req["headers"])
        if "data" in req:
            req["data"] = _strip_pii(req["data"])
        if "cookies" in req:
            req["cookies"] = _REDACTED
        # Strip query strings that may contain tokens
        if "query_string" in req:
            req["query_string"] = _REDACTED

    # Scrub breadcrumb data
    if "breadcrumbs" in event:
        for bc in event.get("breadcrumbs", {}).get("values", []):
            if "data" in bc:
                bc["data"] = _strip_pii(bc["data"])

    # Scrub extra context
    if "extra" in event:
        event["extra"] = _strip_pii(event["extra"])

    # Scrub user info — keep id for correlation but drop everything else
    if "user" in event:
        event["user"] = {"id": event["user"].get("id", "anon")}

    return event


def _before_send_transaction(
    event: dict[str, Any], hint: dict[str, Any]
) -> dict[str, Any] | None:
    """Strip PII from performance transactions."""
    return _before_send(event, hint)


def init_sentry(config: SentryConfig) -> bool:
    """Initialise Sentry SDK if enabled and DSN is configured.

    Returns True if Sentry was successfully initialised.
    """
    if not config.enabled or not config.dsn:
        logger.debug("Sentry disabled or no DSN configured — skipping init")
        return False

    try:
        import sentry_sdk
        from sentry_sdk.integrations.fastapi import FastApiIntegration
        from sentry_sdk.integrations.logging import LoggingIntegration

        sentry_sdk.init(
            dsn=config.dsn,
            environment=config.environment,
            traces_sample_rate=config.traces_sample_rate,
            send_default_pii=False,
            before_send=_before_send,
            before_send_transaction=_before_send_transaction,
            integrations=[
                FastApiIntegration(transaction_style="endpoint"),
                LoggingIntegration(level=logging.WARNING, event_level=logging.ERROR),
            ],
            release=_get_release(),
        )
        logger.info("Sentry initialised (env=%s)", config.environment)
        return True
    except Exception:
        logger.exception("Failed to initialise Sentry SDK")
        return False


def capture_exception(exc: Exception, **extra: Any) -> str | None:
    """Capture an exception to Sentry with optional extra context.

    Returns the Sentry event ID, or None if Sentry is not active.
    """
    try:
        import sentry_sdk

        if sentry_sdk.is_initialized():
            with sentry_sdk.new_scope() as scope:
                for k, v in extra.items():
                    scope.set_extra(k, v)
                return sentry_sdk.capture_exception(exc)
    except Exception:
        logger.debug("Failed to send exception to Sentry", exc_info=True)
    return None


def capture_message(message: str, level: str = "error", **extra: Any) -> str | None:
    """Capture a message to Sentry with optional extra context."""
    try:
        import sentry_sdk

        if sentry_sdk.is_initialized():
            with sentry_sdk.new_scope() as scope:
                for k, v in extra.items():
                    scope.set_extra(k, v)
                return sentry_sdk.capture_message(message, level=level)
    except Exception:
        logger.debug("Failed to send message to Sentry", exc_info=True)
    return None


def _get_release() -> str:
    """Return the ATC version string for Sentry release tracking."""
    try:
        from atc import __version__

        return f"atc@{__version__}"
    except Exception:
        return "atc@unknown"

File: frontend/src/utils/sentry.ts
/**
 * Sentry SDK initialisation and helpers for ATC frontend.
 *
 * Sentry is opt-in — it only activates when the user gives consent and
 * a DSN is provided via the VITE_SENTRY_DSN env var.
 */

/// <reference types="vite/client" />

import * as Sentry from "@sentry/react";

const CONSENT_KEY = "atc:sentry_consent";

/** Whether the user has opted in to crash reporting. */
export function hasConsent(): boolean {
  return localStorage.getItem(CONSENT_KEY) === "true";
}

/** Set the user's crash-reporting consent preference. */
export function setConsent(value: boolean): void {
  localStorage.setItem(CONSENT_KEY, value ? "true" : "false");
  if (!value && Sentry.isInitialized()) {
    Sentry.getCurrentScope().setClient(undefined);
  }
}

/** Initialise Sentry if consent is given and a DSN is available. */
export function initSentry(): void {
  const dsn = import.meta.env.VITE_SENTRY_DSN as string | undefined;
  if (!dsn || !hasConsent()) return;

  Sentry.init({
    dsn,
    environment: import.meta.env.MODE as string,
    sendDefaultPii: false,
    beforeSend(event) {
      return stripPii(event);
    },
  });
}

/** Send a user-initiated report via the backend API. */
export async function sendReport(
  message: string,
  context?: Record<string, unknown>,
): Promise<{ sent: boolean; event_id?: string }> {
  try {
    const res = await fetch("/api/settings/sentry/send-report", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ message, context }),
    });
    if (!res.ok) return { sent: false };
    return res.json();
  } catch {
    return { sent: false };
  }
}

/** Capture a frontend exception to Sentry (client-side). */
export function captureException(
  error: unknown,
  extra?: Record<string, unknown>,
): void {
  if (!Sentry.isInitialized()) return;
  Sentry.withScope((scope) => {
    if (extra) {
      Object.entries(extra).forEach(([k, v]) => scope.setExtra(k, v));
    }
    Sentry.captureException(error);
  });
}

// ---------------------------------------------------------------------------
// PII stripping
// ---------------------------------------------------------------------------

const PII_KEYS =
  /password|secret|token|api.?key|auth|credential|cookie|session.?id|email|phone|ssn|credit.?card|card.?number/i;

function stripPii(event: Sentry.ErrorEvent): Sentry.ErrorEvent {
  if (event.request) {
    if (event.request.headers) {
      event.request.headers = redactHeaders(event.request.headers);
    }
    if (event.request.data) {
      event.request.data = redactObj(
        event.request.data as Record<string, unknown>,
      ) as unknown as Sentry.ErrorEvent["request"] extends { data?: infer D }
        ? D
        : never;
    }
    if (event.request.cookies) {
      event.request.cookies = { _: "[Filtered]" };
    }
    if (event.request.query_string) {
      event.request.query_string = "";
    }
  }

  if (event.extra) {
    event.extra = redactObj(event.extra);
  }

  if (event.user) {
    event.user = { id: event.user.id ?? "anon" };
  }

  return event;
}

function redactHeaders(
  headers: Record<string, string>,
): Record<string, string> {
  const result: Record<string, string> = {};
  for (const [k, v] of Object.entries(headers)) {
    result[k] = PII_KEYS.test(k) ? "[Filtered]" : v;
  }
  return result;
}

function redactObj(obj: unknown): Record<string, unknown> {
  if (typeof obj !== "object" || obj === null) return {};
  const result: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj as Record<string, unknown>)) {
    if (PII_KEYS.test(k)) {
      result[k] = "[Filtered]";
    } else if (typeof v === "object" && v !== null) {
      result[k] = redactObj(v);
    } else {
      result[k] = v;
    }
  }
  return result;
}
